### ISSUE TITLE
"NAN prediction in example ..." errors when passing in close-to-zero values

### ISSUE THREAD

**Issue by gustavla** [JohnLangford/vowpal_wabbit#756]
Consider passing in the following data (`y` being the dependent variable):

```
x0,x1,y
<x>,0.0,-1
1.0,0.0,1
0.0,0.0,-1
0.0,1.0,-1
0.0,0.0,-1
1.0,0.0,-1
0.0,0.0,-1
1.0,0.0,1
0.0,0.0,-1
1.0,1.0,1
```

If you replace `<x>` with `0.0`, it will run just fine and train a reasonable model. However, if its absolute value is in the range between `1e-19` and `1e-45`  (since smaller will be read as exactly zero anyway), it will start to spit out plenty of errors like this:

```
average  since         example        example  current  current  current
loss     last          counter         weight    label  predict features
0.693147 0.693147            1            1.0  -1.0000   0.0000        2
NAN prediction in example 2, forcing 0.000000
0.693147 0.693147            2            2.0   1.0000   0.0000        2
NAN prediction in example 3, forcing 0.000000
NAN prediction in example 4, forcing 0.000000
0.693147 0.693147            4            4.0  -1.0000   0.0000        2
NAN prediction in example 5, forcing 0.000000
NAN prediction in example 6, forcing 0.000000
NAN prediction in example 7, forcing 0.000000
```

As a consequence, it will also return a useless model with all NaN coefficients.

**Comment by JohnLangford** [comment]
This seems to architecture specific, because I cannot reproduce on my 
machine.  As an example, when I feed in:

echo "-1 | 1e-20 0
1 | 1 0" | vw

all is well.

I suspect this has to do with the normalization VW does.  When the 
feature value starts out super-tiny, VW will set the corresponding 
feature value very large.  But, why does this introduce a NAN?  I need 
to reproduce to diagnose, and I can't reproduce.

-John

**Comment by trufanov-nok** [comment]
I was able to reproduce this with
``
$ echo "-1 | x1:1e-23 x2:0.0
1 | x1:1.0 x2:0.0
-1 | x1:0.0 x2:0.0
-1 | x1:0.0 x2:1.0
-1 | x1:0.0 x2:0.0
-1 | x1:1.0 x2:0.0
-1 | x1:0.0 x2:0.0
1 | x1:1.0 x2:0.0
-1 | x1:0.0 x2:0.0
1 | x1:1.0 x2:1.0" | ./vw
For more information use: vw --help
Num weight bits = 18
learning rate = 0.5
initial_t = 0
power_t = 0.5
using no cache
Reading datafile = 
num sources = 1
average  since         example        example  current  current  current
loss     last          counter         weight    label  predict features
1.000000 1.000000            1            1.0  -1.0000   0.0000        2
NAN prediction in example 2, forcing 0.000000
1.000000 1.000000            2            2.0   1.0000   0.0000        2
NAN prediction in example 3, forcing 0.000000
NAN prediction in example 4, forcing 0.000000
1.000000 1.000000            4            4.0  -1.0000   0.0000        2
NAN prediction in example 5, forcing 0.000000
NAN prediction in example 6, forcing 0.000000
NAN prediction in example 7, forcing 0.000000
NAN prediction in example 8, forcing 0.000000
1.000000 1.000000            8            8.0   1.0000   0.0000        2
NAN prediction in example 9, forcing 0.000000
NAN prediction in example 10, forcing 0.000000

finished run
number of examples per pass = 10
passes used = 1
weighted example sum = 10.000000
weighted label sum = -4.000000
average loss = 1.000000
best constant = -0.400000
best constant's loss = 0.840000
total feature number = 17
``
It was fine with 1e-20 - 1e-22.
So it may be tracked down to particular line in learn, but I suspect it couldn't be fixed anyway.

**Comment by JohnLangford** [comment]
Great, I can reproduce as well.

-John

**Comment by JohnLangford** [comment]
I fixed this, at least on my machine. The problem was in the update rule 
and has to do with making updates be unitless and the finite precision 
of floats.

https://github.com/JohnLangford/vowpal_wabbit/blob/master/vowpalwabbit/gd.cc#L440

This should not affect predictions at all, but this will means that 
people using tiny numbers may see slower convergence.

-John

**Comment by arielf** [comment]
I think I can make underflow issues be less common while also speeding up parsing.  This would involve replacing our current (near optimal) "fast parseFloat" with an even better (both faster and more accurate) version.  It is on my TODO list. Hope to get to it this year.

**Comment by arielf** [comment]
John,

I wonder about the choice of 1e-9 as the safety factor in the gd.cc update.

Shouldn't we try to make this safety floor smaller?

32-bit FLT_MIN is 1.17549435E-38F so taking a square root we can't get an underflow in 1/(x*x)  if x is greater than about 1.084202e-19

**Comment by trufanov-nok** [comment]
My imho: I think x shouldn't be modified without need at all whatever small value is added to it. How about this code:
```
weight* w = &fw;
float x2 = x * x;
if (x2 < 1e-39)
 { 
  x = (x>0)?1.084202e-19:-1.084202e-19;
  x2 = 8.507062e-38; // hardcoded 1/(x*x)
}
```

I've got 1-e39 const from
```
   float f = 0.1;
   while (!isinf(1/f)) f/=10;
   qDebug() << f; // f is 1e-39
```

it seems to fix Gustav's example  and passes `make test`.

**Comment by arielf** [comment]
Thanks Alex,
Yes, I think this approach is better.  John?

**Comment by wai7niu8** [comment]
Hi, all:
I meet the same problem when using VW's MF, but the solution above seems can't solve my problem.
There are the params:
cat vw_input | ./vw -b 18 -q ui --rank 10 --l2 3 --learning_rate 0.015 --decay_learning_rate 0.9 --power_t 0 --passes 20 -f vw.model  --cache_file vw.cache
and the result is:
creating quadratic features for pairs: ui
using l2 regularization = 3
final_regressor = vw.model
Num weight bits = 18
learning rate = 0.015
initial_t = 1
power_t = 0
decay_learning_rate = 0.9
using cache_file = vw.cache
ignoring text input in favor of cache input
num sources = 1
average  since         example        example  current  current  current
loss     last          counter         weight    label  predict features
0.667343 0.667343            1            1.0   1.0000   0.1831       23
0.628490 0.589637            2            2.0   1.0000   0.2321       23
0.573036 0.517581            4            4.0   1.0000   0.2994       23
0.840122 1.107208            8            8.0   1.0000   0.3520       23
0.632574 0.425027           16           16.0   1.0000   0.4074       23
3.970465 7.308356           32           32.0   5.0000   0.6162       23
6.304581 8.638697           64           64.0   1.0000   1.1481       23
4.797502 3.290423          128          128.0   2.0000   0.8625       23
2.697726 0.597951          256          256.0   1.0000   0.6555       23
2.137785 1.577844          512          512.0   4.0000   0.7334       23
56.857509 111.577233         1024         1024.0   1.0000   0.5840       23
41.128324 25.399138         2048         2048.0   1.0000   0.5939       23
25.930501 10.732679         4096         4096.0   1.0000   1.1791       23
19.470009 13.009517         8192         8192.0   1.0000   0.7646       23
15.094596 10.719182        16384        16384.0   1.0000   0.7499       23
16.379301 17.664006        32768        32768.0   1.0000   0.5364       23
14.600711 12.822122        65536        65536.0   2.0000   0.6233       23
16.287760 17.974809       131072       131072.0   1.0000   0.7090       23
20.793497 25.299233       262144       262144.0   1.0000   0.7385       23
16.575091 12.356686       524288       524288.0   1.0000   0.6054       23
14.781751 12.988411      1048576      1048576.0   2.0000   0.7033       23
15.769941 16.758130      2097152      2097152.0   1.0000   0.9231       23
14.939491 14.109041      4194304      4194304.0   1.0000   0.7153       23
14.897151 14.854812      8388608      8388608.0   1.0000   0.7733       23
14.918367 14.939582     16777216     16777216.0   1.0000   0.8988       23
13.257977 13.257977     33554432     33554432.0   3.0000   1.8546       23 h
...
...
NAN prediction in example 39410337, forcing 0.000000
NAN prediction in example 39410338, forcing 0.000000
NAN prediction in example 39410339, forcing 0.000000
NAN prediction in example 39410339, forcing 0.000000
NAN prediction in example 39410340, forcing 0.000000
...
...
And it will return a useless model with all NaN coefficients.

**Comment by JohnLangford** [comment]
I agree---this has fewer operations in the standard case.

Note though that changing x here does not change the prediction in any way---it just changes the update rule to be more conservative.

**Comment by arielf** [comment]
Understand the "just changing the update rule to be more conservative", but need it be _so_ conservative as 1e-9?  Using (e.g.) 1.1e-19 should avoid the NaN just as well, while also allowing faster convergence, no?

**Comment by JohnLangford** [comment]
Yes, that's the current:

https://github.com/JohnLangford/vowpal_wabbit/blob/master/vowpalwabbit/gd.cc#L436

-John
